.github/workflows/reusable_publish_version.yml

```diff
@@ -38,12 +37,121 @@ jobs:
+
+      - name: Install a TOML parser
+        run: cargo install --force --locked --version 0.8.1 taplo-cli
+
+      - name: Configure git
+        run: |
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git config user.name "github-actions[bot]"
+          git config --add --bool push.autoSetupRemote true
+
+      - name: Patch release version
+        if: ${{ inputs.environment == 'release' }}
+        run: |
+          set -x
+
+          currentVersion=$(/home/runner/.cargo/bin/taplo get -f lib/Cargo.toml "package.version")
+
+          if [[ $currentVersion == *"-beta"* ]]; then
+            major=$(echo $currentVersion | tr "." "\n" | sed -n 1p)
+            minor=$(echo $currentVersion | tr "." "\n" | sed -n 2p)
+            betaNum=$(echo $currentVersion | tr "." "\n" | sed -n 4p)
+            version=${major}.${minor}.$(($betaNum - 1))
+
+            # Bump the crate version
+            sed -i "s#^version = \".*\"#version = \"${version}\"#" Cargo.toml
+            sed -i "s#^version = \".*\"#version = \"${version}\"#" lib/Cargo.toml
+
+            # Update Cargo.lock without updating dependency versions
+            cargo check --no-default-features --features storage-mem
+
+            # Commit changes
+            git commit -am "Prepare v${version} release"
+          else
+            version=${currentVersion}
+          fi
+
+          # Create the tag
+          git tag -a v${version} -m "Release ${version}"
+
+      - name: Create or patch beta branch
+        if: ${{ inputs.environment == 'beta' }}
+        run: |
+          set -x
+
+          currentVersion=$(/home/runner/.cargo/bin/taplo get -f lib/Cargo.toml "package.version")
+
+          if [[ $currentVersion == *"-beta"* ]]; then
+            major=$(echo $currentVersion | tr "." "\n" | sed -n 1p)
+            minor=$(echo $currentVersion | tr "." "\n" | sed -n 2p)
+            patchAndMeta=$(echo $currentVersion | tr "." "\n" | sed -n 3p)
+            betaNum=$(echo $currentVersion | tr "." "\n" | sed -n 4p)
+            betaVersion=${major}.${minor}.${patchAndMeta}.$(($betaNum + 1))
+          else
+            git checkout -b releases/beta
+            major=$(echo $currentVersion | tr "." "\n" | sed -n 1p)
+            minor=$(echo $currentVersion | tr "." "\n" | sed -n 2p)
+            betaVersion=${major}.$(($minor + 1)).0-beta.1
+          fi
+
+          # Bump the crate version
+          sed -i "s#^version = \".*\"#version = \"${betaVersion}\"#" Cargo.toml
+          sed -i "s#^version = \".*\"#version = \"${betaVersion}\"#" lib/Cargo.toml
+
+          # Update Cargo.lock without updating dependency versions
+          cargo check --no-default-features --features storage-mem
+
+          # Commit changes
+          git commit -am "Prepare v${betaVersion} release"
+
+          # Create the tag
+          git tag -a v${betaVersion} -m "Release ${betaVersion}"
```

Comment:
- They are idempotent for `nightly`.
- `release` is not idempotent because we re-use the beta branch to temporarily store the release commit. We could easily make it idempotent though by introducing `releases/stable` as an intermediary branch.
- `beta` is also not idempotent because it bumps the version when a beta branch already exists. That's so we can also use it to patch existing betas. We could make it idempotent by adding an option to not bump the version.

As you pointed out, tags are another issue currently affecting idempotence but this should be an easy fix for both `beta` and `release` once we fix the above issues. `nightly` doesn't create tags so it's already idempotent.

Having said that, I know crates are idempotent but I'm not sure about Docker images and binaries. Is pushing those idempotent @sgirones?

Reply:
yeah, Docker images and binaries are just overwritten.

Don't treat this comment as a blocker. We can iterate from here 👍

Reply:
Awesome! Thanks for the review @sgirones! I have created https://github.com/surrealdb/surrealdb/issues/3124 for this. Will follow up on it right after I finish what I'm working on now.